### DISCUSSION TITLE
What is the proper way to shutdown a session?

### DISCUSSION THREAD

**Discussion by mangodan2003** [eclipse-zenoh/roadmap#160] [Q&A]
I am using only pub so far.
The basic examples all (zenoh-c and zenonh-cpp) seem to shutdown cleanly but as soon as I try to integrate into a real application I am hitting rust panics or just hangs at exit.

I am using zenoh-c and zenoh-cpp 1.0.0.5

This is the simplest example that reproduces the panic - note, holding the session within a class which is destructed causes the panic. I have tried using pure zenoh-c and calling z_close in the class destructor but thus doesn't help. In fact not calling z_close stops the panic and in a simple example makes everything appear good, but in our real world app something hangs and prevents the app from exiting.

```
#include <memory>
#include "zenoh.hxx"
using namespace zenoh;

const char* keyexpr = "test";

class ZenohPub {
public:
  static ZenohPub& getZenohPub() {
    static ZenohPub instance;
    return instance;
  }

  void send(const char* buf) {
    pub->put(Bytes::serialize(buf), {.encoding = Encoding("text/plain")});
  }

private:
  ZenohPub() {
    auto config = Config::create_default();
    session = std::make_unique<zenoh::Session>(Session::open(std::move(config)));
    pub = std::make_unique<zenoh::Publisher>(session->declare_publisher(KeyExpr(keyexpr)));
  }
  std::unique_ptr<zenoh::Session> session;
  std::unique_ptr<zenoh::Publisher> pub;
};

int main(int argc, char **argv) {
  ZenohPub& zp = ZenohPub::getZenohPub();
  zp.send("test!");
}
```

Build with:
`g++ -o test test.cpp -I zenoh-cpp/include/ -I zenoh-c/include/ -DZENOHCXX_ZENOHC -lzenohc -L zenoh-c/target/release
`

Run with:
`
LD_LIBRARY_PATH=zenoh-c/target/release  ./test
`

The output of which is:
`thread '<unnamed>' panicked at /rustc/129f3b9964af4d4a709d1383930ade12dfe7c081/library/std/src/thread/local.rs:262:26:
cannot access a Thread Local Storage value during or after destruction: AccessError
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
Aborted (core dumped)
`

Many thanks!

**Comment by wyfo** [comment]
The difference between the zenoh examples and your code is the keyword `static`. In fact, according to the [specifications](https://en.cppreference.com/w/cpp/utility/program/exit), static objects are destructed in `atexit`. 
However, the Rust layer behind the C/C++ API doesn't work well in `atexit` context; Rust thread local storage has indeed already been destructed at that time, and `z_close` panics when accessing it.
This is a known issue, and the only solution we can advise for now is to avoid calling `z_close` (or `Session` destructor) in `atexit`, thus avoid static session object.
